service/history/replication/sync_state_retriever.go

```diff
@@ -376,9 +389,40 @@ func (s *SyncStateRetrieverImpl) getSnapshot(mutableState workflow.MutableState)
 	return mutableStateProto, nil
 }
 
-func (s *SyncStateRetrieverImpl) getEventsBlob(ctx context.Context, branchToken []byte, startEventId int64, endEventId int64) ([]*commonpb.DataBlob, error) {
+func (s *SyncStateRetrieverImpl) getEventsBlob(
+	ctx context.Context,
+	workflowKey definition.WorkflowKey,
+	versionHistory *history.VersionHistory,
+	startEventId int64,
+	endEventId int64,
+) ([]*commonpb.DataBlob, error) {
+	var eventBlobs []*commonpb.DataBlob
+
+	if s.eventBlobCache != nil {
+		for {
+			eventVersion, err := versionhistory.GetVersionHistoryEventVersion(versionHistory, startEventId)
+			if err != nil {
+				return nil, err
+			}
+			xdcCacheValue, ok := s.eventBlobCache.Get(persistence.NewXDCCacheKey(workflowKey, startEventId, eventVersion))
+			if !ok {
+				break
+			}
+			left := endEventId - startEventId
+			if startEventId != common.FirstEventID && int64(len(xdcCacheValue.EventBlobs)) >= left {
```

Comment:
hmm just realized startEventID can also be 1 for a normal workflow.
Maybe we should add the nextEventID for the newRun to raw task as well. We can refactor later, it's just a sanity check.